myia/info.py

```diff
@@ -0,0 +1,103 @@
+"""Objects and routines to track debug information."""
+
+from typing import Any
+import types
+import threading
+import traceback
+
+
+# We use per-thread storage for the about stack.
+_about = threading.local()
+_about.stack = [None]
+
+
+def current_info():
+    """Return the `DebugInfo` for the current context."""
+    return _about.stack[-1]
+
+
+class DebugInfo(types.SimpleNamespace):
+    """Debug information for an object.
+
+    Attributes:
+        about: DebugInfo of a different object, that this
+            object is derived from in some way
+        relation: How the object relates to `about`
+        save_trace: Whether the trace of the DebugInfo's
+            creation should be saved.
+        trace: The trace at the moment of the DebugInfo's
+            creation, or None.
+
+    """
+
+    def __init__(self, **kwargs):
+        """Construct a DebugInfo object."""
+        self.about = None
+        self.relation = None
+        self.save_trace: bool = False
+        self.trace: Any = None
+        top = current_info()
+        if top:
+            # Only need to look at the top of the stack
+            self.__dict__.update(top.__dict__)
+
+        self._id: int = None
+        self.__dict__.update(kwargs)
+
+        if self.save_trace:
+            # We remove the last entry that corresponds to
+            # this line in the code.
+            self.trace = traceback.extract_stack()[:-1]
+
+    def __enter__(self):
+        """Set this `DebugInfo` as a template in this context.
+
+        Any `DebugInfo` created within the context of
+        `with self: ...` will inherit all attributes of `self`.
+        """
+        _about.stack.append(self)
+
+    def __exit__(self, type, value, tb):
+        """Exit the context of this `DebugInfo`."""
+        assert _about.stack[-1] is self
+        _about.stack.pop()
+
+
+class AutoNamedDebugInfo(DebugInfo):
```

Comment:
`NamedDebugInfo`? They can be manually named after all, and it's shorter.

Reply:
Sure.